### PR TITLE
feat: allow themes to have arbitrary tokens

### PR DESCRIPTION
**Pull request by dbanksdesign**
#### Description of changes

Tries to open up the theme object a bit so people can add whatever design tokens they want, while keeping _some_ type safety and intellisense. 

#### Description of how you validated changes

https://github.com/aws-amplify/amplify-ui/assets/321279/c758f6c7-e8a2-45aa-927e-8301b4e51a15

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] `yarn test` passes and tests are updated/added
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/ui/src/theme/tokens/space.ts [hunk 1]
```diff
@@ -70,3 +73,6 @@ export const space: Space<'default'> = {
     full: { value: '100%' },
   },
 };
+
+// I want to be able to pass in a Theme object that has extra tokens
+// and it returns that same object type WITH the extra tokens
```
Comment (calebpollman) [inline packages/ui/src/theme/tokens/space.ts:78]: Oh do you? 😅

File: packages/ui/src/theme/tokens/types/designToken.ts [hunk 2]
```diff
@@ -5,6 +5,34 @@ import { Properties } from 'csstype';
  */
 export type DesignToken<ValueType = unknown> = { value: ValueType } | ValueType;
 
+/**
+ * Opening up our theme to accept design tokens we don't explicitly define.
+ * This helper type allows you to have a deeply nested object where the nodes are DesignTokens
+ */
```
Comment (calebpollman) [inline packages/ui/src/theme/tokens/types/designToken.ts:11]: Nit: remove first person verbiage and reasoning from documentation:
```suggestion
/**
 * Helper type util allowing creation of a deeply nested object of Design Tokens
 */
```

File: packages/ui/src/theme/tokens/types/designToken.ts [hunk 3]
```diff
@@ -5,6 +5,34 @@ import { Properties } from 'csstype';
...
+export type RecursiveDesignToken<
```
Comment (calebpollman) [inline packages/ui/src/theme/tokens/types/designToken.ts:12]: Sort of annoying but can `RecursiveDesignToken` and `DesignTokenValue` be moved to the bottom of the file with the other type utils?

File: packages/ui/src/theme/tokens/types/designToken.ts [hunk 4]
```diff
@@ -5,6 +5,34 @@ import { Properties } from 'csstype';
...
+  ValueType = unknown,
+  Output extends OutputVariantKey = unknown,
+  Platform extends PlatformKey = unknown
+> = {
+  [key: string | number]:
+    | DesignTokenValue<ValueType, Output, Platform>
+    | RecursiveDesignToken<ValueType, Output, Platform>;
```
Comment (calebpollman) [inline packages/ui/src/theme/tokens/types/designToken.ts:19]: How do we prevent infinite recursion here?

File: packages/ui/src/theme/tokens/types/designToken.ts [hunk 5]
```diff
@@ -5,6 +5,34 @@ import { Properties } from 'csstype';
...
+/**
+ * Opening up our theme to accept design tokens we don't explicitly define.
+ * This helper type allows you to have a deeply nested object where the nodes are DesignTokens
+ */
+export type RecursiveDesignToken<
+  ValueType = unknown,
+  Output extends OutputVariantKey = unknown,
+  Platform extends PlatformKey = unknown
+> = {
+  [key: string | number]:
```
Comment (calebpollman) [inline packages/ui/src/theme/tokens/types/designToken.ts:17]: Think we can use the `PropKey` union later in this file here

File: packages/ui/src/theme/tokens/space.ts (hunk 1 above)
Comment (dbanksdesign) [reply packages/ui/src/theme/tokens/space.ts:78]: lol whoops

**Review by reesscot — APPROVED** [review]

**Review by calebpollman — APPROVED** [review]

**Pull request merged**